### PR TITLE
feat(lambda-event-sources): support for batching window to sqs event source

### LINKED ISSUE #11722: (aws-lambda-event-sources): add batching window support for SQS event sources
add support for feature launch: https://aws.amazon.com/about-aws/whats-new/2020/11/aws-lambda-now-supports-batch-windows-of-up-to-5-minutes-for-functions/
* customer can define batching window up to 300 seconds
* if batching window is defined batch size can be up to 10000
* if batching window is not defined batch size can be up to 10

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### LINKED ISSUE #13770: @aws-cdk/aws-lambda-event-sources: Update the batch size and enable the batch window options
### Use Case

The `SqsEventSource` in `@aws-cdk/aws-lambda-event-sources` allows for a `batchSize` between 1 and 10.

As of November 24, 2020 it is now possible to specify a batch size of up to 10,000, and a batching window of 5 minutes.

This is helpful to increase throughput and efficiency when using Lambda with SQS.

### Proposed Solution

The `SqsEventSource` should be modified to allow for these larger batch sizes and allow specifying a batching window.

```js
fn.addEventSource(
  new SqsEventSource(queue, {
    batchSize: 1000,
    maximumBatchingWindow: Duration.minutes(1)
  }),
);
```

### Other

Here's the announcement on the new changes:
https://aws.amazon.com/about-aws/whats-new/2020/11/aws-lambda-now-supports-batch-windows-of-up-to-5-minutes-for-functions/

Here is where this has been updated for AWS SAM:
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-sqs.html

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by stefanodesjo**
Fixes #11722

This is continued work from #11724

closes #13770

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by stefanodesjo** [comment]
@nija-at @MrArnoldPalmer Anything remaining to get an approval?

**Comment by nija-at** [comment]
Hello - 

Sorry about the delay in reviewing PRs. We are experiencing an increased backlog of items that need our attention.
We will get to your PR as soon as we are able to. Apologies for the delay.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 1]
```diff
@@ -14,6 +14,16 @@ export interface SqsEventSourceProps {
...
 
+  /**
+   * The maximum amount of time to gather records before invoking the function.
+   * Maximum of Duration.minutes(5)
+   *
+   * Valid Range: Minimum value of 0 minutes. Maximum value of 5 minutes.
+   *
+   * @default - no batching window. The lambda function will be invoked immediately with the records that are available.
+   */
+  readonly maxBatchingWindow?: Duration;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:25]: Just `batchingWindow` is sufficient. Analogous to `batchSize`.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 2]
```diff
@@ -14,6 +14,16 @@ export interface SqsEventSourceProps {
    */
   readonly batchSize?: number;
 
+  /**
+   * The maximum amount of time to gather records before invoking the function.
+   * Maximum of Duration.minutes(5)
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:19]: Captured in the next line already.

```suggestion
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 3]
```diff
@@ -29,14 +39,29 @@ export class SqsEventSource implements lambda.IEventSource {
...
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
-    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
-      throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+    const maxBatchingWindow = this.props.maxBatchingWindow;
+    if (maxBatchingWindow !== undefined) {
+      if (queue.fifo) {
+        throw new Error('Batching window is not supported for FIFO queues');
+      }
+      if (maxBatchingWindow.toSeconds() > 300) {
+        throw new Error(`Maximum batching window must be 300 seconds or less (given ${maxBatchingWindow.toSeconds()})`);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:48]:
```suggestion
        throw new Error(`Maximum batching window must be 300 seconds or less (given ${maxBatchingWindow.toHumanString()})`);
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 4]
```diff
@@ -29,14 +39,29 @@ export class SqsEventSource implements lambda.IEventSource {
...
+      }
+    }
+    if (this.props.batchSize !== undefined && !Token.isUnresolved(this.props.batchSize)) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:51]: Glad you're checking for tokens here.

`maxBatchingWindow` can also be a token. You can check this by calling `maxBatchingWindow.isUnresolved()`.

File: packages/@aws-cdk/aws-lambda-event-sources/README.md [hunk 5]
```diff
@@ -67,6 +67,7 @@ const queue = new sqs.Queue(this, 'MyQueue', {
 
 lambda.addEventSource(new SqsEventSource(queue, {
   batchSize: 10, // default
+  maxBatchingWindow: Duration.minutes(5),
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/README.md:70]: Can you add a couple of sentences briefly describing batching (with a link for further details) above this code snippet?

**Review by nija-at — DISMISSED** [review]
Hi @stefanodesjo - Thanks for submitting this PR.

See some of my comments below.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:25]: Looks like we're calling it `maxBatchingWindow` elsewhere already - https://github.com/aws/aws-cdk/blob/b9407102304f043adcd9a4fc1cde4d23d3da9004/packages/%40aws-cdk/aws-lambda-event-sources/lib/stream.ts#L73-L79.

Ignore this comment. It's fine to leave it as is.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 6]
```diff
@@ -29,14 +38,35 @@ export class SqsEventSource implements lambda.IEventSource {
...
+      if (queue.fifo) {
+        throw new Error('Batching window is not supported for FIFO queues');
+      }
+      if (maxBatchingWindow.toSeconds() > 300) {
+        throw new Error(`Maximum batching window must be 300 seconds or less (given ${maxBatchingWindow.toHumanString()})`);
+      }
+    }
+    if (this.props.batchSize !== undefined && !Token.isUnresolved(this.props.batchSize)) {
+      if (
+        this.props.maxBatchingWindow !== undefined && !Token.isUnresolved(this.props.maxBatchingWindow) &&
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:52]: I'm not sure I understand why you're checking for token here? 
Whether it's a token or not, as long as it's specified, max batch size should be between 1 and 10000, no?

(my previous comment was referring to the conditional at line-46)

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 7]
```diff
@@ -29,14 +38,35 @@ export class SqsEventSource implements lambda.IEventSource {
...
+        (this.props.batchSize < 1 || this.props.batchSize > 10000)
+      ) {
+        throw new Error(`Maximum batch size must be between 1 and 10000 inclusive (given ${this.props.batchSize}) when batching window is specified.`);
+      }
+      if (
+        (this.props.maxBatchingWindow === undefined || Token.isUnresolved(this.props.maxBatchingWindow)) &&
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:58]: same comment as above.

**Review by nija-at — DISMISSED** [review]
Hi @stefanodesjo -

Looks like you misunderstood my previous comments. See below.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts (hunk 6 above)
Comment (stefanodesjo) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts]: Yes, misunderstanding :)

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 8]
```diff
@@ -29,14 +38,28 @@ export class SqsEventSource implements lambda.IEventSource {
   private _eventSourceMappingId?: string = undefined;
 
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
-    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
-      throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+    if (this.props.maxBatchingWindow !== undefined && !Token.isUnresolved(this.props.maxBatchingWindow)) {
+      if (queue.fifo) {
+        throw new Error('Batching window is not supported for FIFO queues');
+      }
+      if (this.props.maxBatchingWindow.toSeconds() > 300) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:45]: These checks are still not precise. You will need to change it this way -

```suggestion
    if (this.props.maxBatchingWindow !== undefined) {
      if (queue.fifo) {
        throw new Error('Batching window is not supported for FIFO queues');
      }
      if (!this.props.maxBatchingWindow.isUnresolved() && this.props.maxBatchingWindow.toSeconds() > 300) {
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 9]
```diff
@@ -85,6 +85,38 @@ export = {
...
+      'EventSourceArn': {
+        'Fn::GetAtt': [
+          'Q63C6E3AB',
+          'Arn',
+        ],
+      },
+      'FunctionName': {
+        'Ref': 'Fn9270CBC0',
+      },
+    }));
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:115]: The assertion should check that the `BatchSize` property is set to 500

```suggestion
    // THEN
    expect(stack).to(haveResource('AWS::Lambda::EventSourceMapping', {
      BatchSize: 500
    }));
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 10]
```diff
@@ -113,6 +145,109 @@ export = {
...
+    expect(stack).to(haveResource('AWS::Lambda::EventSourceMapping', {
+      'EventSourceArn': {
+        'Fn::GetAtt': [
+          'Q63C6E3AB',
+          'Arn',
+        ],
+      },
+      'FunctionName': {
+        'Ref': 'Fn9270CBC0',
+      },
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:170]: It's sufficient to just check for the properties the test is concerned about.

```suggestion
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 11]
```diff
@@ -113,6 +145,109 @@ export = {
...
+    // GIVEN
+    const stack = new cdk.Stack();
+    const fn = new TestFunction(stack, 'Fn');
+    const q = new sqs.Queue(stack, 'Q');
+
+    // WHEN/THEN
+    test.throws(() => fn.addEventSource(new sources.SqsEventSource(q, {
+      batchSize: 11000,
+      maxBatchingWindow: cdk.Duration.minutes(5),
+    })), /Maximum batch size must be between 1 and 10000 inclusive \(given 11000\)/);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:188]:
```suggestion
    })), /Maximum batch size must be between 1 and 10000/i);
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 12]
```diff
@@ -113,6 +145,109 @@ export = {
...
+    expect(stack).to(haveResource('AWS::Lambda::EventSourceMapping', {
+      'EventSourceArn': {
+        'Fn::GetAtt': [
+          'Q63C6E3AB',
+          'Arn',
+        ],
+      },
+      'FunctionName': {
+        'Ref': 'Fn9270CBC0',
+      },
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:214]: same as above comment.

```suggestion
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 13]
```diff
@@ -113,6 +145,109 @@ export = {
...
+  'fails if batch window is > 5'(test: Test) {
+    // GIVEN
+    const stack = new cdk.Stack();
+    const fn = new TestFunction(stack, 'Fn');
+    const q = new sqs.Queue(stack, 'Q');
+
+    // WHEN/THEN
+    test.throws(() => fn.addEventSource(new sources.SqsEventSource(q, {
+      maxBatchingWindow: cdk.Duration.minutes(7),
+    })), /Maximum batching window must be 300 seconds or less \(given 7 minutes\)/);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:246]:
```suggestion
    })), /Maximum batching window must be 300 seconds or less/i);
```

**Review by nija-at — DISMISSED** [review]
Thanks for updating the PR. See my comments below.

Could you also look at the failing build and fix it? The logs to the failing build are available in the automated comment above.

**Review by nija-at — APPROVED** [review]

**Comment by mckalexee** [comment]
@stefanodesjo @nija-at 

Thanks for the quick turnaround here! 🙏

I really appreciate it.

**Pull request merged**